File: src/apps/asteroids.cpp
// [Revision: v1.1] [Path: src/apps/asteroids.cpp] [Date: 2025-12-10]
// Description: Asteroids. Tuned for 60Hz loop. Standardized Game Over font.

#include "asteroids.h"

// Physics Constants (Tuned for 60 FPS)
const float DRAG = 0.985;      // Slightly less drag to maintain momentum
const float ACCEL = 0.15;      // Stronger thrust for responsiveness
const float ROT_SPEED = 0.12;  // Faster rotation
const float MAX_VEL = 4.0;     // Higher top speed
const int SCREEN_W = 128;
const int SCREEN_H = 64;

AsteroidsApp::AsteroidsApp() {
    bullets.resize(5);   
    particles.resize(20); 
}

void AsteroidsApp::start() {
    u8g2.setContrast(systemContrast);
    resetGame();
}

void AsteroidsApp::stop() {
    asteroids.clear();
}

void AsteroidsApp::resetGame() {
    score = 0;
    level = 1;
    gameOver = false;
    shipPos = {SCREEN_W / 2.0f, SCREEN_H / 2.0f};
    shipVel = {0, 0};
    shipAngle = -PI / 2; 
    
    asteroids.clear();
    for(int i=0; i<3; i++) {
        spawnAsteroid(random(0, SCREEN_W), random(0, SCREEN_H), 3);
    }
}

void AsteroidsApp::spawnAsteroid(float x, float y, int size) {
    Asteroid a;
    a.pos = {x, y};
    float speed = (4 - size) * 0.4; // Adjusted speed scale
    float angle = random(0, 628) / 100.0;
    a.vel = {cos(angle) * speed, sin(angle) * speed};
    a.size = size;
    a.radius = size * 3 + 2; 
    a.seed = random(0, 100);
    a.active = true;
    asteroids.push_back(a);
}

void AsteroidsApp::spawnBullet() {
    for(auto &b : bullets) {
        if(!b.active) {
            b.active = true;
            b.pos = shipPos;
            b.vel.x = shipVel.x + cos(shipAngle) * 5.0; // Faster bullets
            b.vel.y = shipVel.y + sin(shipAngle) * 5.0;
            b.life = 50; // Last slightly longer (approx 0.8s)
            break;
        }
    }
}

void AsteroidsApp::createExplosion(float x, float y, int count) {
    int spawned = 0;
    for(auto &p : particles) {
        if(!p.active && spawned < count) {
            p.active = true;
            p.pos = {x, y};
            float angle = random(0, 628) / 100.0;
            float speed = random(50, 250) / 100.0;
            p.vel = {cos(angle) * speed, sin(angle) * speed};
            p.life = random(15, 30); // ~0.5s at 60fps
            spawned++;
        }
    }
}

// --------------------------------------------------------------------------
// INPUT & UPDATE
// --------------------------------------------------------------------------

void AsteroidsApp::handleInput(char key) {
    if (gameOver && key == '5') {
        resetGame();
        return;
    }
    if (!gameOver && key == '5') {
        spawnBullet();
    }
}

void AsteroidsApp::update() {
    if (gameOver) return;

    // 1. POLL INPUT
    bool left = false, right = false, thrust = false;
    for(int i=0; i<activeKeyCount; i++) {
        if (activeKeys[i] == '4') left = true;
        if (activeKeys[i] == '6') right = true;
        if (activeKeys[i] == '2') thrust = true;
    }

    // 2. SHIP PHYSICS
    if (left) shipAngle -= ROT_SPEED;
    if (right) shipAngle += ROT_SPEED;
    
    isThrusting = thrust;
    if (thrust) {
        shipVel.x += cos(shipAngle) * ACCEL;
        shipVel.y += sin(shipAngle) * ACCEL;
    }
    
    // Drag
    shipVel.x *= DRAG;
    shipVel.y *= DRAG;
    
    // Cap Speed (Manual clamp to keep vector direction correct roughly)
    // Simple clamp per axis is okay for this retro feel, or normalize vector
    float speed = sqrt(shipVel.x*shipVel.x + shipVel.y*shipVel.y);
    if (speed > MAX_VEL) {
        shipVel.x = (shipVel.x / speed) * MAX_VEL;
        shipVel.y = (shipVel.y / speed) * MAX_VEL;
    }
    
    shipPos.x += shipVel.x;
    shipPos.y += shipVel.y;

    // Wrap
    if (shipPos.x < 0) shipPos.x += SCREEN_W;
    if (shipPos.x > SCREEN_W) shipPos.x -= SCREEN_W;
    if (shipPos.y < 0) shipPos.y += SCREEN_H;
    if (shipPos.y > SCREEN_H) shipPos.y -= SCREEN_H;

    // 3. BULLETS
    for(auto &b : bullets) {
        if (b.active) {
            b.pos.x += b.vel.x;
            b.pos.y += b.vel.y;
            b.life--;
            
            if (b.pos.x < 0) b.pos.x += SCREEN_W;
            if (b.pos.x > SCREEN_W) b.pos.x -= SCREEN_W;
            if (b.pos.y < 0) b.pos.y += SCREEN_H;
            if (b.pos.y > SCREEN_H) b.pos.y -= SCREEN_H;

            if (b.life <= 0) b.active = false;
        }
    }

    // 4. PARTICLES
    for(auto &p : particles) {
        if (p.active) {
            p.pos.x += p.vel.x;
            p.pos.y += p.vel.y;
            p.life--;
            if (p.life <= 0) p.active = false;
        }
    }

    // 5. ASTEROIDS & COLLISIONS
    bool levelCleared = true;
    
    for(auto &a : asteroids) {
        if (!a.active) continue;
        levelCleared = false;
        
        a.pos.x += a.vel.x;
        a.pos.y += a.vel.y;
        
        if (a.pos.x < 0) a.pos.x += SCREEN_W;
        if (a.pos.x > SCREEN_W) a.pos.x -= SCREEN_W;
        if (a.pos.y < 0) a.pos.y += SCREEN_H;
        if (a.pos.y > SCREEN_H) a.pos.y -= SCREEN_H;

        // Collision: Ship
        if (checkCollision(shipPos.x, shipPos.y, 3, a.pos.x, a.pos.y, a.radius)) {
            createExplosion(shipPos.x, shipPos.y, 10);
            gameOver = true;
        }

        // Collision: Bullets
        for(auto &b : bullets) {
            if (b.active && checkCollision(b.pos.x, b.pos.y, 1, a.pos.x, a.pos.y, a.radius)) {
                b.active = false;
                a.active = false;
                createExplosion(a.pos.x, a.pos.y, 5);
                score += 100 / a.size;
                if (a.size > 1) {
                    spawnAsteroid(a.pos.x, a.pos.y, a.size - 1);
                    spawnAsteroid(a.pos.x, a.pos.y, a.size - 1);
                }
                break;
            }
        }
    }

    if (levelCleared) {
        level++;
        shipPos = {SCREEN_W/2.0f, SCREEN_H/2.0f};
        shipVel = {0,0};
        for(int i=0; i<level+2; i++) {
            spawnAsteroid(random(0,SCREEN_W), random(0,SCREEN_H), 3);
        }
    }
}

bool AsteroidsApp::checkCollision(float x1, float y1, float r1, float x2, float y2, float r2) {
    float dx = x1 - x2;
    float dy = y1 - y2;
    float distSq = dx*dx + dy*dy;
    float rSum = r1 + r2;
    return distSq < (rSum * rSum);
}

void AsteroidsApp::drawWireframeModel(const float* coords, int numPoints, float x, float y, float angle, float scale) {
    float cosA = cos(angle);
    float sinA = sin(angle);
    
    int x0, y0, x1, y1;
    
    auto transform = [&](int idx, int &outX, int &outY) {
        float px = coords[idx*2] * scale;
        float py = coords[idx*2+1] * scale;
        outX = x + (px * cosA - py * sinA);
        outY = y + (px * sinA + py * cosA);
    };

    transform(0, x0, y0);
    for(int i=1; i<=numPoints; i++) {
        int idx = i % numPoints; 
        transform(idx, x1, y1);
        u8g2.drawLine(x0, y0, x1, y1);
        x0 = x1;
        y0 = y1;
    }
}

void AsteroidsApp::render() {
    if (gameOver) {
        // Standardized Game Over Screen
        u8g2.setFont(u8g2_font_ncenB12_tr);
        u8g2.drawStr(8, 30, "GAME OVER");
        
        char buf[20];
        sprintf(buf, "SCORE: %d", score);
        u8g2.setFont(FONT_SMALL);
        u8g2.drawStr(30, 50, buf);
        u8g2.drawStr(10, 62, "PRESS 5 TO RESTART");
        return;
    }

    // Ship
    const float shipModel[] = { 4,0, -3,-3, -2,0, -3,3 }; 
    drawWireframeModel(shipModel, 4, shipPos.x, shipPos.y, shipAngle, 1.0);
    
    // Thrust
    if (isThrusting) {
        u8g2.drawLine(
            shipPos.x - cos(shipAngle)*5, 
            shipPos.y - sin(shipAngle)*5,
            shipPos.x - cos(shipAngle)*2, 
            shipPos.y - sin(shipAngle)*2
        );
    }

    // Bullets
    for(auto &b : bullets) {
        if(b.active) u8g2.drawDisc(b.pos.x, b.pos.y, 2);
    }

    // Particles
    for(auto &p : particles) {
        if(p.active) u8g2.drawPixel(p.pos.x, p.pos.y);
    }

    // Asteroids
    for(auto &a : asteroids) {
        if(a.active) {
            u8g2.drawDisc(a.pos.x, a.pos.y, a.radius);
        }
    }

    // UI
    char scoreBuf[16];
    sprintf(scoreBuf, "%d", score);
    u8g2.setFont(u8g2_font_micro_tr); 
    u8g2.drawStr(2, 6, scoreBuf);
    u8g2.drawFrame(0, 0, 128, 64);

}

File: data/asteroids.lua
-- asteroids.lua - Classic Asteroids Game
-- Controls: 4=Left, 6=Right, 2=Thrust, 5=Fire, 5=Restart(gameover)

-- Constants
local SCREEN_W, SCREEN_H = gfx.width(), gfx.height()
local DRAG = 0.98
local ACCEL = 0.15
local ROT_SPEED = 0.12
local MAX_VEL = 4.0
local PI = 3.14159

-- Game state
local shipX, shipY = SCREEN_W / 2, SCREEN_H / 2
local shipVX, shipVY = 0, 0
local shipAngle = -PI / 2
local isThrusting = false
local gameOver = false
local score = 0
local level = 1

-- Bullets
local bullets = {}
local MAX_BULLETS = 5

-- Asteroids
local asteroids = {}

-- Particles
local particles = {}

-- Helper: wrap position
local function wrap(val, max)
    if val < 0 then return val + max end
    if val > max then return val - max end
    return val
end

-- Helper: distance squared
local function distSq(x1, y1, x2, y2)
    local dx, dy = x1 - x2, y1 - y2
    return dx * dx + dy * dy
end

-- Spawn asteroid
local function spawnAsteroid(x, y, size)
    local speed = (4 - size) * 0.5
    local angle = math.random() * 2 * PI
    table.insert(asteroids, {
        x = x, y = y,
        vx = math.cos(angle) * speed,
        vy = math.sin(angle) * speed,
        size = size,
        radius = size * 3 + 2,
        active = true
    })
end

-- Create explosion particles
local function explode(x, y, count)
    for i = 1, count do
        local angle = math.random() * 2 * PI
        local speed = math.random() * 2 + 0.5
        table.insert(particles, {
            x = x, y = y,
            vx = math.cos(angle) * speed,
            vy = math.sin(angle) * speed,
            life = math.random(10, 25),
            active = true
        })
    end
end

-- Reset game
local function resetGame()
    shipX, shipY = SCREEN_W / 2, SCREEN_H / 2
    shipVX, shipVY = 0, 0
    shipAngle = -PI / 2
    gameOver = false
    score = 0
    level = 1
    
    bullets = {}
    asteroids = {}
    particles = {}
    
    -- Spawn initial asteroids
    for i = 1, 3 do
        spawnAsteroid(math.random(0, SCREEN_W), math.random(0, SCREEN_H), 3)
    end
end

-- Fire bullet
local function fireBullet()
    if #bullets >= MAX_BULLETS then return end
    table.insert(bullets, {
        x = shipX,
        y = shipY,
        vx = shipVX + math.cos(shipAngle) * 5,
        vy = shipVY + math.sin(shipAngle) * 5,
        life = 50,
        active = true
    })
end

-- Handle input
local function handleInput()
    input.scan()
    
    if gameOver then
        if input.pressed("5") then
            resetGame()
        end
        return
    end
    
    -- Rotation
    if input.held("4") then shipAngle = shipAngle - ROT_SPEED end
    if input.held("6") then shipAngle = shipAngle + ROT_SPEED end
    
    -- Thrust
    isThrusting = input.held("2")
    if isThrusting then
        shipVX = shipVX + math.cos(shipAngle) * ACCEL
        shipVY = shipVY + math.sin(shipAngle) * ACCEL
    end
    
    -- Fire
    if input.pressed("5") then
        fireBullet()
    end
end

-- Update game
local function updateGame()
    if gameOver then return end
    
    -- Ship physics
    shipVX = shipVX * DRAG
    shipVY = shipVY * DRAG
    
    -- Speed cap
    local speed = math.sqrt(shipVX * shipVX + shipVY * shipVY)
    if speed > MAX_VEL then
        shipVX = (shipVX / speed) * MAX_VEL
        shipVY = (shipVY / speed) * MAX_VEL
    end
    
    shipX = wrap(shipX + shipVX, SCREEN_W)
    shipY = wrap(shipY + shipVY, SCREEN_H)
    
    -- Update bullets
    for i = #bullets, 1, -1 do
        local b = bullets[i]
        if b.active then
            b.x = wrap(b.x + b.vx, SCREEN_W)
            b.y = wrap(b.y + b.vy, SCREEN_H)
            b.life = b.life - 1
            if b.life <= 0 then
                table.remove(bullets, i)
            end
        end
    end
    
    -- Update particles
    for i = #particles, 1, -1 do
        local p = particles[i]
        if p.active then
            p.x = p.x + p.vx
            p.y = p.y + p.vy
            p.life = p.life - 1
            if p.life <= 0 then
                table.remove(particles, i)
            end
        end
    end
    
    -- Update asteroids and check collisions
    local levelCleared = true
    for i = #asteroids, 1, -1 do
        local a = asteroids[i]
        if a.active then
            levelCleared = false
            a.x = wrap(a.x + a.vx, SCREEN_W)
            a.y = wrap(a.y + a.vy, SCREEN_H)
            
            -- Ship collision
            if distSq(shipX, shipY, a.x, a.y) < (3 + a.radius) * (3 + a.radius) then
                explode(shipX, shipY, 10)
                gameOver = true
            end
            
            -- Bullet collisions
            for j = #bullets, 1, -1 do
                local b = bullets[j]
                if b.active and distSq(b.x, b.y, a.x, a.y) < a.radius * a.radius then
                    b.active = false
                    a.active = false
                    explode(a.x, a.y, 5)
                    score = score + math.floor(100 / a.size)
                    
                    -- Split asteroid
                    if a.size > 1 then
                        spawnAsteroid(a.x, a.y, a.size - 1)
                        spawnAsteroid(a.x, a.y, a.size - 1)
                    end
                    
                    table.remove(asteroids, i)
                    table.remove(bullets, j)
                    break
                end
            end
        end
    end
    
    -- Next level
    if levelCleared and not gameOver then
        level = level + 1
        shipX, shipY = SCREEN_W / 2, SCREEN_H / 2
        shipVX, shipVY = 0, 0
        for i = 1, level + 2 do
            spawnAsteroid(math.random(0, SCREEN_W), math.random(0, SCREEN_H), 3)
        end
    end
end

-- Draw ship
local function drawShip()
    local cos_a, sin_a = math.cos(shipAngle), math.sin(shipAngle)
    
    -- Ship triangle points
    local nose_x = shipX + cos_a * 5
    local nose_y = shipY + sin_a * 5
    local left_x = shipX + math.cos(shipAngle + 2.5) * 4
    local left_y = shipY + math.sin(shipAngle + 2.5) * 4
    local right_x = shipX + math.cos(shipAngle - 2.5) * 4
    local right_y = shipY + math.sin(shipAngle - 2.5) * 4
    
    gfx.line(math.floor(nose_x), math.floor(nose_y), math.floor(left_x), math.floor(left_y))
    gfx.line(math.floor(left_x), math.floor(left_y), math.floor(right_x), math.floor(right_y))
    gfx.line(math.floor(right_x), math.floor(right_y), math.floor(nose_x), math.floor(nose_y))
    
    -- Thrust flame
    if isThrusting then
        local flame_x = shipX - cos_a * 6
        local flame_y = shipY - sin_a * 6
        gfx.line(math.floor(shipX - cos_a * 3), math.floor(shipY - sin_a * 3),
                 math.floor(flame_x), math.floor(flame_y))
    end
end

-- Draw game
local function drawGame()
    gfx.clear()
    
    if gameOver then
        gfx.setFont(2)
        gfx.text(15, 25, "GAME OVER")
        gfx.setFont(0)
        gfx.text(35, 40, "Score: " .. score)
        gfx.text(20, 55, "Press 5 to restart")
    else
        -- Border
        gfx.rect(0, 0, SCREEN_W, SCREEN_H)
        
        -- Ship
        drawShip()
        
        -- Bullets
        for _, b in ipairs(bullets) do
            if b.active then
                gfx.fillCircle(math.floor(b.x), math.floor(b.y), 1)
            end
        end
        
        -- Particles
        for _, p in ipairs(particles) do
            if p.active then
                gfx.pixel(math.floor(p.x), math.floor(p.y))
            end
        end
        
        -- Asteroids
        for _, a in ipairs(asteroids) do
            if a.active then
                gfx.fillCircle(math.floor(a.x), math.floor(a.y), a.radius)
            end
        end
        
        -- Score
        gfx.setFont(0)
        gfx.text(2, 7, tostring(score))
    end
    
    gfx.send()
end

-- Main loop
print("Asteroids started - Press * or # to exit")
resetGame()

local running = true
while running do
    handleInput()
    updateGame()
    drawGame()
    
    if input.held("*") or input.held("#") then
        running = false
    end
    
    sys.delay(16)
    sys.yield()
end

print("Asteroids ended. Final score: " .. score)
